backend/src/main/scala/bloop/io/ClasspathHasher.scala

```diff
@@ -1,284 +1,231 @@
-                  }
-                }
-              }
+        val entryPromise = Promise[FileHash]()
+        val promise = Option(hashingPromises.putIfAbsent(entry, entryPromise))
+
+        promise match {
+          // The hashing is done by this process, compute hash and complete the promise
+          case None =>
+            Computed(makeHashingTask(entry, entryPromise))
+          // The hashing is acquired by another process, wait on its result
+          case Some(promise) =>
+            logger.debug(s"Wait for hashing of $entry to complete")
+            FromPromise(waitForAnotherTask(entry, promise))
+        }
+      }
+    }
+
+    def waitForAnotherTask(entry: Path, promise: Promise[FileHash]): Task[FileHash] =
+      Task.fromFuture(promise.future).flatMap { hash =>
+        hash.hash() match {
+          case BloopStamps.cancelledHash =>
+            if (cancelCompilation.isCompleted)
+              Task.now(FileHash.of(entry, BloopStamps.cancelledHash))
+            else {
+              // If the process that acquired it cancels the computation, try acquiring it again
+              logger
+                .warn(s"Unexpected hash computation of $entry was cancelled, restarting...")
+              acquireHashingEntry(entry).task
             }
-          )
-          found
+          case _ =>
+            Task.now(hash)
         }
+      }
 
-        val attrs = Files.readAttributes(jar.toPath, classOf[BasicFileAttributes])
-        val currentMetadata = (FileTime.fromMillis(IO.getModifiedTimeOrZero(jar)), attrs.size())
+    val onCancel = Task(isCancelled.compareAndSet(false, true)).unit
 
-        Option(definedMacrosJarCache.get(jar)) match {
-          case Some((metadata, hit)) if metadata == currentMetadata => jar -> hit
-          case _ =>
-            val detected = detectMacro(jar)
-            definedMacrosJarCache.put(jar, (currentMetadata, detected))
-            jar -> detected
+    val (computing, fromAnotherTask) =
+      classpath.toVector.map(path => acquireHashingEntry(path.underlying)).partition {
+        case _: Computed => true
+        case _: FromPromise => false
+      }
+
+    // first compute all the entries that are computed by this run
+    // then start waiting for entries that are computed by another tasks
+    val hashes = for {
+      computed <- Task.parSequenceN(parallelUnits)(computing.map(_.task).toList)
+      fromPromises <- Task.parSequenceN(parallelUnits)(fromAnotherTask.map(_.task).toList)
+    } yield computed ++ fromPromises
+
+    hashes
+      .map { result =>
+        if (isCancelled.get || cancelCompilation.isCompleted) {
+          cancelCompilation.trySuccess(())
+          Left(())
+        } else {
+          Right(result.toVector)
         }
       }
-    }
+      .doOnCancel(onCancel)
+  }
+}
 
-    Task.gatherUnordered(classpath.map(readJar(_)))
+object ClasspathHasher {
+  final val global: ClasspathHasher = new ClasspathHasher
```

Comment:
It was impossible to test `ClasspathHasher` because tests were sharing global cache. I turned it into class and create `global` instance because:
- I didn't want to touch already existing code which was referring to `ClasspathHasher` as an object. Now it's using `ClasspathHasher.global` which is almost the same
- in tests, each testcase is creating a fresh instance of `ClasspathHasher` - no pollution